src/DeviceInput/deviceInputSystem.ts

```diff
@@ -0,0 +1,310 @@
+        }
+
+        // Pointer Events
+        if (this._mouseActive || this._touchActive) {
+            this._elementToAttachTo.removeEventListener("pointermove", this._pointerMoveEvent);
+            this._elementToAttachTo.removeEventListener("pointerdown", this._pointerDownEvent);
+            this._elementToAttachTo.removeEventListener("pointerup", this._pointerUpEvent);
+        }
+
+        // Gamepad Events
+        window.removeEventListener("gamepadconnected", this._gamepadConnectedEvent);
+        window.removeEventListener("gamepaddisconnected", this._gamepadDisconnectedEvent);
+    }
+
+    // Private functions
+    /**
+     * Add device and inputs to device map
+     * @param deviceName Assigned name of device (may be SN)
+     * @param numberOfInputs Number of input entries to create for given device
+     */
+    private _registerDevice(deviceName: string, numberOfInputs: number) {
+        if (!this._inputs[deviceName]) {
+            const device = new Array<number>(numberOfInputs);
+
+            for (let i = 0; i < numberOfInputs; i++) {
+                device[i] = 0;
+            }
+
+            this._inputs[deviceName] = device;
+            this.onDeviceConnectedObservable.notifyObservers(deviceName);
+        }
+    }
+
+    /**
+     * Given a specific device name, remove that device from the device map
+     * @param deviceName Name of device to be removed
+     */
+    private _unregisterDevice(deviceName: string) {
+        if (this._inputs[deviceName]) {
+            delete this._inputs[deviceName];
+            this.onDeviceDisconnectedObservable.notifyObservers(deviceName);
+        }
+    }
+
+    /**
+     * Handle all actions that come from keyboard interaction
+     */
+    private _handleKeyActions() {
+        this._keyboardDownEvent = ((evt) => {
+            if (!this._keyboardActive) {
+                this._keyboardActive = true;
+                this._registerDevice(DeviceInputSystem.KEYBOARD_DEVICE, DeviceInputSystem._MAX_KEYCODES);
+            }
+
+            const kbKey = this._inputs[DeviceInputSystem.KEYBOARD_DEVICE];
+            if (kbKey) {
+                kbKey[evt.keyCode] = 1;
+            }
+        });
+
+        this._keyboardUpEvent = ((evt) => {
+            const kbKey = this._inputs[DeviceInputSystem.KEYBOARD_DEVICE];
+            if (kbKey) {
+                kbKey[evt.keyCode] = 0;
+            }
+        });
+
+        window.addEventListener("keydown", this._keyboardDownEvent);
+        window.addEventListener("keyup", this._keyboardUpEvent);
+    }
+
+    /**
+     * Handle all actions that come from pointer interaction
+     */
+    private _handlePointerActions() {
+        this._pointerMoveEvent = ((evt) => {
+            if (evt.pointerType == "mouse") {
+                if (!this._mouseActive) {
+                    this._mouseActive = true;
+                    this._registerDevice(DeviceInputSystem.MOUSE_DEVICE, DeviceInputSystem._MAX_MOUSE_INPUTS);

```

Comment:
I still struggle to understand why we are not using evt.pointerId instead of DeviceInputSystem._MAX_MOUSE_INPUTS
same for the pointer below.

Keyboard can be at -1  and gamepads from -2 to the infinite (and beyond)

Reply:
Overall we should not treat the mouse differently than the touch (pointer events are meant to handle that). We should have:
- pointers
- keyboard
- gamepads

Reply:
When I tested using pointerId for the mouse, all of the pointerIds were the same, regardless of the mouse click.  I chose to go with the MAX_MOUSE_INPUTS approach because it leveraged the MouseEvent API to be able to distinguish between the different types of clicks (via MouseEvent.button).

Reply:
"When I tested using pointerId for the mouse, all of the pointerIds were the same, regardless of the mouse click": this is what we want as we need to identify the mouse as an unique entity

Reply:
Talked offline, combined mouse/touch into singular pointer handling